include/boost/rational.hpp

```diff
@@ -94,18 +99,34 @@ namespace boost {
 template <typename IntType>
 IntType gcd(IntType n, IntType m)
 {
-    // Defer to the version in Boost.Math
+    // Defer to the version in Boost.Integer
     return integer::gcd( n, m );
 }
 
 template <typename IntType>
 IntType lcm(IntType n, IntType m)
 {
-    // Defer to the version in Boost.Math
+    // Defer to the version in Boost.Integer
     return integer::lcm( n, m );
 }
 #endif  // BOOST_CONTROL_RATIONAL_HAS_GCD
 
+namespace rational_detail{
+
+   template <class FromInt, class ToInt>
+   struct is_compatible_integer
+   {
+      BOOST_STATIC_CONSTANT(bool, value = ((std::numeric_limits<FromInt>::is_specialized && std::numeric_limits<FromInt>::is_integer
+         && (std::numeric_limits<FromInt>::digits <= std::numeric_limits<ToInt>::digits)
+         && (std::numeric_limits<FromInt>::radix == std::numeric_limits<ToInt>::radix)
+         && (std::numeric_limits<FromInt>::is_signed == std::numeric_limits<ToInt>::is_signed)
```

Comment:
This forbids conversion from a smaller unsigned type to a larger signed type, which is safe.